sound/soc/soc-compress.c

```diff
@@ -72,10 +73,20 @@ static int soc_compr_components_free(struct snd_compr_stream *cstream,
 static int soc_compr_open(struct snd_compr_stream *cstream)
 {
 	struct snd_soc_pcm_runtime *rtd = cstream->private_data;
-	struct snd_soc_component *component;
+	struct snd_soc_component *component, *save = NULL;
+	struct snd_soc_rtdcom_list *rtdcom;
 	struct snd_soc_dai *cpu_dai = rtd->cpu_dai;
 	int ret;
 
+	for_each_rtd_components(rtd, rtdcom, component) {
+		ret = pm_runtime_get_sync(component->dev);
+		if (ret < 0 && ret != -EACCES) {
+			pm_runtime_put_noidle(component->dev);
+			save = component;
```

Comment:
@crojewsk what is the rationale here? What are we "saving" here? and the pm_runtime calls are unbalanced: get_sync->put_noidle? why?

Reply:
This is based on review from @plbossart for the exact same patch. @ranj063 please note that these patches were simply separated from probes PR and pushed here. There are no changes in between. All the reviews from probe PR still apply.
Judging by the rest of the code in kernel, I do believe Pierre is right about the ref leak.

Reply:
@crojewsk Sorry I got confused a bit but I still think there's a problem here. Assuming the get_sync() failed halfway in the for_each_rtd_components(), you only save the component that failed but call mark_last_busy() and put_autosuspend() for all components including the ones that would come after the saved one. Is that correct?

Reply:
And how to I know the order of the items within the collections stays the same - ain't there a chance some other thread alters the list? If that is not the case, I could update the exit part.

Reply:
I dont know if the list would get altered or not and I dont think we can make any assumptions about it. But the logic currently would definitely be unbalanced if you quit the loop when you encounter the first comp that failed but go on the suspend all of them anyway

Reply:
Then how would you see it done? Right now the only other option seems to be caching all the failures (or successes?) within a list or array and either skipping all of them within the closing scope in case of failures, or looping only through the successful ones.

Reply:
this is a tricky one. soc_pcm_open() doesnt care about the return value for get_sync(). Not sure if thats the right thing to so.

BTW the other thing I noticed is the unbalanced mutex_unlock() when you goto out from line 86.

Reply:
I think the problem might be the "goto out". Maybe we need to get rid of it? and let the actions below fail for the component that couldnt be resumed?

Reply:
I agree with Pierre here - there are many places in kernel where this ain't handled properly. soc_pcm_open is one of these, unfortunately. Even spoke with Rafael about this during ELC-E. He too agreed about the existence of the problem - though this is left for the caller to fix : )

The unbalanced mutex is noted. Indeed after adding goto, the rest of the code hasn't been updated accordingly. Thanks!
I'm keen to agree on remove-goto idea. This would destine the failing device to fail yet again, but without the certainty of collection being kept in check, I do not see many good choices here.

Reply:
Hmm. If I remove goto, then 'save' is useless and for the failing device we got 2x put == unbalanced ref count.

Slowly coming down to conclusion that my initial patch was the right one - based on soc_pcm_open and without any handling. Here we are dealing with collection of devices, not a single device and to my understanding one cannot be sure about list preserving it's state until the function gets to the exit scope.

Reply:
@crojewsk I also feel the same. The check for get_sync() is just complicating things unnecessarily here

Reply:
@plbossart What is your opinion on this? I see the comment you made below, but I do not wish to ignore this conversation in the process. Are we all agreeing on revert-to-status-quo solution -or- the 'break' being added in the closing scope? Again, the later assumes the collection does not alter throughout the entire lifespan of the function.

I'm fine with going with either of these to alsa-devel and awaiting community's review.

Reply:
I am fine with leaving out the error handling on get_sync() for now as long as we have a follow-up set of patches to fix this for both PCM and compressed stuff (should be the same solution really).

Reply:
Aligned with 'break' suggestion and pushed the changes in v2. The rest is for community to decide.